### PR TITLE
Refactor stub methods

### PR DESCRIPTION
**Pull request by mroderick**
This PR refactors `sinon.stub()` and `sandbox.stub()` (internally `collection.stub()` methods to improve legibility.

Feedback on naming of files and functions as well as locations of these in the tree is most welcome.

The aim was to not make *any* changes to functionality, but make it easier to understand the `stub` methods and to highlight differences between them for future improvements.

During refactoring this, I discovered an inconsistency in the API.

#### Proposal: fix inconsistency of stubbing non-function properties

As you can see in `collection.stub` and its tests, it supports stubbing non-function properties, which `sinon.stub` does not.

The documentation for `sandbox.stub` reads:

> The sandbox `stub` method can also be used to stub any kind of property. This is useful if you need to override an object's property for the duration of a test, and have it restored when the test completes

In my opinion, stubs should be used for replacing methods and objects with fakes. Instead of stubbing out a property on an object for the duration of a test, the author should use a stub in the right place to return an entirely fake object that has the (fake) methods and properties that is needed for the test.

I propose that we remove stubbing of non-function properties feature from `collection.stub`, so that it's behaviour matches that of `sinon.stub` (since it would then defer everything to `sinon.stub`).

I am happy to do that as part of this PR.

What do you think?

### CODE REVIEW

File: lib/sinon/collection.js [hunk 1]
```diff
@@ -91,46 +90,22 @@ var collection = {
...
-                    typeof getPropertyDescriptor(propOwner, prop).value === "function" &&
-                    stubbedObj.hasOwnProperty(prop)
-                ) {
-                    col.add(stubbedObj[prop]);
-                }
-            });
-
-            return stubbedObj;
+        if (isStubbingEntireObject) {
+            collectOwnMethods(stubbed).forEach(this.add.bind(this));
```
Comment (mantoni) [inline lib/sinon/collection.js]: Not that it matters here, but you know that you can pass a custom `this` to `forEach` as the second argument?

**Comment by mantoni** [comment]
This is quite a bit of cleanup! I like how you made things more readable by moving the complex decision making logic into separate modules. Looks all good to me.

Regarding the API inconsistency you spotted: I remember running into this as well a while back, but I think the only reason it wasn't cleaned up was laziness. It makes sense to fold this into the stub implementation instead of a collection-only feature.

File: lib/sinon/collection.js (hunk 1 above)
Comment (mroderick) [inline lib/sinon/collection.js]: I did, don't use it much, as I try to avoid using `this` as I find it reduces problems when I do :)

Comment (mantoni) [inline lib/sinon/collection.js]: Absolutely, but since you're using `this` here already ... 😆

**Comment by mroderick** [comment]
> It makes sense to fold this into the stub implementation instead of a collection-only feature.

My proposal was to remove it from `collection.stub`. Are you proposing that we add it to `sinon.stub` instead?

**Comment by mantoni** [comment]
Oh, I didn't read careful enough then. I now remember that it's only implemented for sandboxes because there wouldn't be a possibility to restore the previous value, because the `restore` function is a property of the stub. I think it's a nice feature. We should keep it.

File: lib/sinon/collection.js [hunk 2]
```diff
@@ -11,12 +11,11 @@
 var sinonSpy = require("./spy");
 var sinonStub = require("./stub");
 var sinonMock = require("./mock");
-var walk = require("./util/core/walk");
-var getPropertyDescriptor = require("./util/core/get-property-descriptor");
-var valueToString = require("./util/core/value-to-string");
+var failOnFalsyObject = require("./fail-on-falsy-object");
```
Comment (dazld) [inline lib/sinon/collection.js]: how about `throwOnFalsyObject` - as this tells you explicitly what this will do. `fail` could mean many different things.

Comment (mroderick) [inline lib/sinon/collection.js]: I agree, that is a better name. I've renamed the file, the function and all the variables in new commits.

Thank you for the feedback 👍

File: lib/sinon/stub-descriptor.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+var spy = require("./spy");
+var wrapMethod = require("./util/core/wrap-method");
+
+// This is deprecated and will be removed in a future version of sinon.
+// We will only consider pull requests that fix serious bugs in the implementation
+function stubDescriptor(object, property, descriptor) {
+    var wrapper;
+
+    deprecated.printWarning(
+      "sinon.stub(obj, 'meth', fn) is deprecated and will be removed from" +
```
Comment (lucasfcosta) [inline lib/sinon/stub-descriptor.js]: Hi, friends, I noticed that the third argument for the stub function is no longer going to be valid. While this seems to be okay when stubbing functions I'd like to know how could we stub the `get/set` properties of descriptors.

Are we going to have a specific API for that or users should just use `Object.getOwnPropertyDescriptor` and then stub the `get` and `set` functions in there?

Also, if you want to take a look at how it works right now, #692 is the PR responsible for introducing it.

Comment (fatso83) [inline lib/sinon/stub-descriptor.js]: Good call, although I am not sure.

Comment (mroderick) [inline lib/sinon/stub-descriptor.js]: @lucasfcosta that's an interesting discussion. I am not sure it should block merging of this PR. If you agree, would you mind opening a separate issue for the discussion, so we can merge this?

Comment (lucasfcosta) [inline lib/sinon/stub-descriptor.js]: Yeah, I totally agree.

Let's discuss that in a separate issue, I've got some ideas for the API and how to implement it, I'd love to help with this. Also, this PR LG(reat)TM.

Awesome work, sir 😄

**Comment by mroderick** [comment]
OK, since there are no objections, I am merging this

**Pull request merged**